crates/red_knot_python_semantic/src/semantic_index/definition.rs

```diff
@@ -454,7 +490,10 @@ impl DefinitionKind<'_> {
             | DefinitionKind::Comprehension(_)
             | DefinitionKind::WithItem(_)
             | DefinitionKind::MatchPattern(_)
-            | DefinitionKind::ExceptHandler(_) => DefinitionCategory::Binding,
+            | DefinitionKind::ExceptHandler(_)
+            | DefinitionKind::TypeVar(_)
+            | DefinitionKind::ParamSpec(_)
+            | DefinitionKind::TypeVarTuple(_) => DefinitionCategory::Binding,
```

Comment:
I don't think making them declarations would have any effect.

They are "declared" in their own scope, and then used in the function or class body scope, which is a nested scope. There's no prohibition against redefining a declared name from an outer scope; you're just shadowing it with your own local, which is allowed. So making them declarations doesn't actually prevent you from arbitrarily shadowing the typevar name with something totally different inside the function/class body.

The only thing it would prevent is reassigning the name _in the type param scope_. But this isn't possible anyway, because type param scopes can't contain statements or named expressions.

Later we'll also want to support typevars of the form `T = typing.TypeVar("T")`. But these won't even be `DefinitionKind::TypeVar`, they'll just be `DefinitionKind::Assignment`. And since we won't know until type inference that the RHS even is a typevar, I don't think we could special-case those to be declarations even if we wanted to (and I don't think we do; I think declaration-or-not should be determined purely by the syntactic form, not by the type of the RHS of an assignment.)

Reply:
(I did end up deciding to make them declarations anyway -- even if it makes no difference, I think it makes semantic sense, and actually `DeclarationAndBinding` is marginally more efficient since it doesn't have to look for other reaching declarations.)